File: src/main/scala/net/machinemuse/numina/network/MusePacket.scala
package net.machinemuse.numina.network

import java.io._
import java.util.zip.GZIPOutputStream

import io.netty.buffer.{ByteBufOutputStream, Unpooled}
import net.machinemuse.numina.general.MuseLogger
import net.minecraft.client.entity.EntityPlayerSP
import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.item.ItemStack
import net.minecraft.nbt.{CompressedStreamTools, NBTTagCompound}
import net.minecraft.network.PacketBuffer
import net.minecraftforge.fml.common.network.internal.FMLProxyPacket
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

/**
  * Author: MachineMuse (Claire Semple)
  * Created: 12:58 AM, 09/05/13
  */
abstract class MusePacket {
  val packager: MusePackager

  def write()

  val bytes = new PacketBuffer(Unpooled.buffer())
  val dataout = new DataOutputStream(new ByteBufOutputStream(bytes))

  /**
    * Gets the MC packet associated with this MusePacket
    *
    * @return Packet250CustomPayload
    */
  def getFMLProxyPacket: FMLProxyPacket = {
    dataout.writeInt(MusePacketHandler.packagers.inverse.get(packager).get)
    write()
    new FMLProxyPacket(bytes, MusePacketHandler.networkChannelName)
  }

  def getPacket131 = this

  /**
    * Called by the network manager since it does all the packet mapping
    *
    * @param player
    */
  @SideOnly(Side.CLIENT)
  def handleClient(player: EntityPlayerSP) {}

  def handleServer(player: EntityPlayerMP) {}

  def writeInt(i: Int) {
    dataout.writeInt(i)
  }


  def writeIntArray(data: Array[Int]) {
    try {
      dataout.writeInt(data.length)
      for (k <- data.indices) {
        dataout.writeInt(data(k))
      }
    } catch {
      case e: IOException => {
        e.printStackTrace()
      }
    }
  }

  def writeBoolean(b: Boolean) {
    try {
      dataout.writeBoolean(b)
    } catch {
      case e: IOException => e.printStackTrace()
    }
  }

  def writeDouble(i: Double) {
    try {
      dataout.writeDouble(i)
    } catch {
      case e: IOException => e.printStackTrace()
    }
  }

  /**
    * Writes the IC2ItemTest's ID (short), then size (byte), then damage. (short)
    */
  def writeItemStack(stack: ItemStack) {
    try {
      if (stack == null) {
        dataout.writeShort(-1)
      }
      else {
        val nbt = new NBTTagCompound
        stack.writeToNBT(nbt)
        writeNBTTagCompound(nbt)
      }
    } catch {
      case e: IOException => e.printStackTrace()
    }
  }

  /**
    * Writes a compressed NBTTagCompound to the OutputStream
    */
  protected def writeNBTTagCompound(nbt: NBTTagCompound) {
    if (nbt == null) {
      dataout.writeShort(-1)
    } else {
      val compressednbt: Array[Byte] = compress(nbt)
      dataout.writeShort(compressednbt.length.toShort)
      dataout.write(compressednbt)
    }
  }

  /**
    * Writes a String to the DataOutputStream
    */
  def writeString(string: String) {
    try {
      dataout.writeShort(string.length)
      dataout.writeChars(string)
    } catch {
      case e: IOException => e.printStackTrace()
    }
  }

  /*
   * "Borrowed" from 1.7.10
   */
  @throws(classOf[IOException])
  def compress(p_74798_0_ : NBTTagCompound) : Array[Byte] = {
    val bytearrayoutputstream : ByteArrayOutputStream = new ByteArrayOutputStream
    val dataoutputstream : DataOutputStream = new DataOutputStream(new GZIPOutputStream(bytearrayoutputstream))
    try {
      CompressedStreamTools.write(p_74798_0_, dataoutputstream)
    } finally {
      dataoutputstream.close
    }
    return bytearrayoutputstream.toByteArray
  }
}

trait MusePackager {
  val READ_ERROR: Short = -150

  import net.machinemuse.numina.network.RichInputStream._

  def read(datain: DataInputStream, player: EntityPlayer): MusePacket

  def readByte(datain: DataInputStream): Byte = datain.readByte

  def readShort(datain: DataInputStream): Short = datain.readShort

  def readInt(datain: DataInputStream): Int = datain.readInt

  def readLong(datain: DataInputStream): Long = datain.readLong

  def readBoolean(datain: DataInputStream): Boolean = datain.readBoolean

  def readFloat(datain: DataInputStream): Float = datain.readFloat

  def readDouble(datain: DataInputStream): Double = datain.readDouble

  def readIntArray(datain: DataInputStream): Array[Int] = datain.readIntArray

  def readString(datain: DataInputStream): String = datain.readString

  def readItemStack(datain: DataInputStream): ItemStack = datain.readItemStack

  def readNBTTagCompound(datain: DataInputStream): NBTTagCompound = datain.readNBTTagCompound

  def safeRead[T](codec: () => T): Option[T] = {
    try {
      Some(codec.apply())
    } catch {
      case e: IOException => MuseLogger.logException("PROBLEM READING DATA FROM PACKET D:", e)
    }
  }
}

File: src/main/java/net/machinemuse/numina/network/MusePacket.java
package net.machinemuse.numina.network;

import cpw.mods.fml.common.network.internal.FMLProxyPacket;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufOutputStream;
import io.netty.buffer.Unpooled;
import net.machinemuse.numina.general.MuseLogger;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompressedStreamTools;
import net.minecraft.nbt.NBTTagCompound;

import java.io.DataOutputStream;
import java.io.IOException;

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 12:58 AM, 09/05/13
 *
 * Ported to Java by lehjr on 10/25/16.
 */
public abstract class MusePacket {
    private final ByteBuf bytes;
    private final DataOutputStream dataout;

    public abstract MusePackager packager();

    public abstract void write();

    public ByteBuf bytes() {
        return this.bytes;
    }

    public DataOutputStream dataout() {
        return this.dataout;
    }

    public FMLProxyPacket getFMLProxyPacket() throws IOException {
        this.dataout.writeInt((Integer) MusePacketHandler.packagers.inverse().get(this.packager()));
        this.write();
        return new FMLProxyPacket(this.bytes, MusePacketHandler.networkChannelName);
    }

    public MusePacket getPacket131() {
        return this;
    }

    @SideOnly(Side.CLIENT)
    public void handleClient(EntityPlayer player) {
    }

    public void handleServer(EntityPlayerMP player) {
    }

    public void writeInt(int i) {
        try {
            this.dataout.writeInt(i);
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public void writeIntArray(int[] data) {
        try {
            this.dataout.writeInt(data.length);
            for (int k :  data)
                dataout.writeInt(k);
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public void writeBoolean(boolean b) {
        try {
            this.dataout.writeBoolean(b);
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public void writeDouble(double i) {
        try {
            this.dataout.writeDouble(i);
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public void writeItemStack(ItemStack stack) {
        try {
            if (stack == null) {
                this.dataout.writeShort(-1);
            }
            else {
                NBTTagCompound nbt = new NBTTagCompound();
                stack.writeToNBT(nbt);
                this.writeNBTTagCompound(nbt);
            }
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public void writeNBTTagCompound(NBTTagCompound nbt) {
        try {
            if (nbt == null) {
                this.dataout.writeShort(-1);
            }
            else {
                byte[] compressednbt = CompressedStreamTools.compress(nbt);
                this.dataout.writeShort((short)compressednbt.length);
                this.dataout.write(compressednbt);
            }
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public void writeString(String string) {
        try {
            this.dataout.writeUTF(string);
        } catch (IOException exception) {
            MuseLogger.logException("PROBLEM WRITING DATA TO PACKET:", exception);
        }
    }

    public MusePacket() {
        this.bytes = Unpooled.buffer();
        this.dataout = new DataOutputStream(new ByteBufOutputStream(this.bytes));
    }
}